crates/aranya-daemon/src/sync/task/quic/psk.rs

```diff
@@ -0,0 +1,238 @@
+use buggy::BugExt as _;
+use s2n_quic::provider::tls::rustls::rustls::{
+    client,
+    crypto::{hash::HashAlgorithm, PresharedKey},
+    pki_types::ServerName,
+    server,
+};
+use tokio::sync::mpsc;
+use tracing::{error, warn};
+
+use crate::{keystore::LocalStore, CE, CS, KS};
+
+pub(crate) type TeamIdPSKPair = (TeamId, Arc<PresharedKey>);
+
+const QUIC_SYNC_PSK_CONTEXT: &[u8] = b"AranyaQuicSync-v1";
+
+#[derive(Clone)]
+pub(crate) struct PskSeed(pub(crate) aranya_crypto::tls::PskSeed<CS>);
+
+impl PskSeed {
+    pub(crate) fn new(rng: &mut impl Csprng, team: TeamId) -> Self {
+        let group = GroupId::from(team.into_id());
+        Self(aranya_crypto::tls::PskSeed::new(rng, &group))
+    }
+
+    pub(crate) fn import_from_ikm(ikm: &[u8; 32], team: TeamId) -> Self {
+        let group = GroupId::from(team.into_id());
+        Self(aranya_crypto::tls::PskSeed::import_from_ikm(ikm, &group))
+    }
+
+    pub(crate) fn load(
+        eng: &mut CE,
+        store: &LocalStore<KS>,
+        id: &PskSeedId,
+    ) -> Result<Option<Self>> {
+        store
+            .get_key(eng, id.into_id())
+            .map(|r| r.map(Self))
+            .map_err(Into::into)
+    }
+
+    pub(crate) fn id(&self) -> Result<PskSeedId, IdError> {
+        self.0.id()
+    }
+
+    pub(crate) fn into_inner(self) -> aranya_crypto::tls::PskSeed<CS> {
+        self.0
+    }
+
+    pub(crate) fn generate_psks(self, team: TeamId) -> impl Iterator<Item = Result<PresharedKey>> {
+        let group = GroupId::from(team.into_id());
+        let policy = PolicyId::default();
+        self.0
+            .generate_psks(
+                QUIC_SYNC_PSK_CONTEXT,
+                group,
+                policy,
+                CipherSuiteId::all().iter().copied(),
+            )
+            .map(|r| {
+                let psk = r?;
+                psk_to_rustls(psk)
+            })
+    }
+}
+
+fn psk_to_rustls(psk: aranya_crypto::tls::Psk<CS>) -> Result<PresharedKey> {
+    let identity = psk.identity().as_bytes();
+    let secret = psk.raw_secret_bytes();
+    let alg = match psk.identity().cipher_suite() {
+        CipherSuiteId::TlsAes128GcmSha256 => HashAlgorithm::SHA256,
+        CipherSuiteId::TlsAes256GcmSha384 => HashAlgorithm::SHA384,
+        CipherSuiteId::TlsChaCha20Poly1305Sha256 => HashAlgorithm::SHA256,
+        CipherSuiteId::TlsAes128CcmSha256 => HashAlgorithm::SHA256,
+        CipherSuiteId::TlsAes128Ccm8Sha256 => HashAlgorithm::SHA256,
+        cs => bail!("unknown ciphersuite {cs}"),
+    };
+    let psk = PresharedKey::external(identity, secret)
+        .context("unable to create PSK")?
+        .with_hash_alg(alg)
+        .context("Invalid hash algorithm")?;
+    Ok(psk)
+}
+
+/// PSK store that's shared between [`super::Syncer`]
+/// and [`super::Server`]
+#[derive(Debug)]
+pub struct PskStore {
+    inner: SyncMutex<PskStoreInner>,
+    // Optional sender to report the selected team
+    active_team_tx: mpsc::Sender<TeamId>,
```

Comment:
I forgot to update this

```suggestion
    // sender to report the selected team
    active_team_tx: mpsc::Sender<TeamId>,
```